File: typescript/graph/graph.ts
type Vertex = [number, number];
type VertexList = Vertex[];

interface IGraph {
  n: number;
  directed: boolean;
  vertices: VertexList[];

  getVerticesNum(): number;
  isDirected(): boolean;
  getVertices(u: number): VertexList;
  isEdge(u: number, v: number): boolean;
  // TODO: isEdge(u: number, v: number, w: number): boolean;
  // TODO: getEdgeWeight(u: number, v: number): number;
  addEdge(u: number, v: number, w?: number): void;
  // TODO: removeEdge(u: number, v: number): boolean;
}

class Graph implements IGraph {
  n: number;
  directed: boolean;
  vertices: VertexList[];

  constructor(n: number, directed?: boolean) {
    this.n = n;
    this.directed = directed ?? false;
    this.vertices = Array(n).fill([]);
  }

  getVerticesNum(): number {
    return this.n;
  }

  getVertices(u: number): VertexList {
    return this.vertices[u];
  }

  isDirected(): boolean {
    return this.directed;
  }

  isEdge(u: number, v: number): boolean {
    if (u > this.n || v > this.n) {
      return false;
    }

    return this.vertices[u].some(([v_, w_]) => v === v_);
  }

  addEdge(u: number, v: number, w: number = 1): void {
    if (u > this.n || v > this.n) {
      throw new Error("Graph.addEdge given vertex larger than graph size");
    }

    const edge: Vertex = [v, w];
    this.vertices[u] = [edge, ...this.vertices[u]];

    if (!this.directed) {
      const edgeR: Vertex = [u, w];
      this.vertices[v] = [edgeR, ...this.vertices[v]];
    }
  }
}

export default Graph;

File: go/graph/graph.go
package graph

import (
	"errors"
	"math"
)

type Vertex struct {
	v      int
	weight int
}

type VertexList = []Vertex

type Graph struct {
	n        int
	directed bool
	vertices []VertexList
}

func NewGraph(n int, directed bool) Graph {
	var vertices = make([]VertexList, n)
	return Graph{n: n, directed: directed, vertices: vertices}
}

func (g *Graph) getVertices(u int) VertexList {
	return g.vertices[u]
}

func find(vertices VertexList, v int) (*Vertex, *int) {
	for i, vertex := range vertices {
		if vertex.v == v {
			return &vertex, &i
		}
	}

	return nil, nil
}

func (g *Graph) getEdgeWeight(u int, v int) int {
	vertex, _ := find(g.getVertices(u), v)
	if vertex != nil {
		return vertex.weight
	}

	return math.MinInt32
}

func some(vertices VertexList, v int) bool {
	for _, vertex := range vertices {
		if vertex.v == v {
			return true
		}
	}

	return false
}

func (g *Graph) isEdge(u int, v int) bool {
	if u > g.n || v > g.n {
		return false
	}

	return some(g.getVertices(u), v)
}

func (g *Graph) addEdge(u int, v int, w int) error {
	if u > g.n || v > g.n {
		return errors.New("Graph.addEdge given vertex larger than graph size")
	}

	edge := Vertex{v: v, weight: w}
	g.vertices[u] = append(g.getVertices(u), edge)

	if !g.directed {
		edgeR := Vertex{v: u, weight: w}
		g.vertices[v] = append(g.getVertices(v), edgeR)
	}

	return nil
}

func remove(vertex VertexList, i int) *VertexList {
	vertex = append(vertex[:i], vertex[i+1:]...)

	return &vertex
}

func (g *Graph) removeEdge(u int, v int) (bool, error) {
	vertex, index := find(g.getVertices(u), v)
	if vertex == nil {
		return false, errors.New("not found")
	}

	g.vertices[u] = *remove(g.getVertices(u), *index)

	return true, nil
}
